### PR TITLE
Support of gcHostOrigin-gcTargetEnv

### PR DESCRIPTION
**Pull request by cadgerfeast**
Adds support for `gcHostOrigin`/`gcTargetEnv` direct values in **ClientApp** constructor.

It also updates the remaining samples to use the new `gcHostOriginQueryParam`/`gcTargetEnvQueryParam` or `gcHostOrigin`/`gcTargetEnv` combination.

### CODE REVIEW

File: src/index.ts [hunk 1]
```diff
@@ -151,30 +159,53 @@ class ClientApp {
...
-     *   pcEnvironmentQueryParam: 'pcEnvironment'
+     *   gcHostOriginQueryParam: 'gcHostOrigin',
+     *   gcTargetEnvQueryParam: 'gcTargetEnv'
      * });
      * ```
      *
      * @param cfg - Runtime config of the client
      */
     constructor(cfg: {
-        /** Name of a query param to auto parse into the pcEnvironment; Must be valid and a single param.  Best Practice. */
```
Comment (justinaray) [inline src/index.ts:161]: nice catch!

File: src/index.ts [hunk 2]
```diff
@@ -151,30 +159,53 @@ class ClientApp {
...
+         */
         gcTargetEnvQueryParam?: string;
+        /**
+         * The GC host origin (e.g. https://apps.mypurecloud.com, https://apps.mypurecloud.au).
+         * Must be a valid origin and used with gcTargetEnv.
+         */
+        gcHostOrigin?: string;
+        /**
+         * The GC env target (e.g. prod, prod-euw2).
+         * Must be a valid origin and used with gcHostOrigin.
```
Comment (justinaray) [inline src/index.ts:195]: nit, `Must be a valid environment name`

File: examples/utils/oauth.js [hunk 3]
```diff
@@ -1,5 +1,5 @@
 // Authenticate with Genesys Cloud
-function authenticate(client, pcEnvironment) {
+function authenticate(client, pcEnvironment, state) {
```
Comment (justinaray) [inline examples/utils/oauth.js:2]: Just need to assert all the calls are updated here.
Also, need to make sure we deploy these utils in all instances.  I do think it's the case, but, want to confirm.

File: examples/agentInteractionsDemo/app.js [hunk 4]
```diff
@@ -152,19 +152,27 @@ new Vue({
...
-            });
-            Vue.prototype.$clientApp = clientApp;
+            if (queryParameters.gcHostOrigin && queryParameters.gcTargetEnv) {
+                clientApp = new window.purecloud.apps.ClientApp({
+                    gcHostOrigin: queryParameters.gcHostOrigin,
+                    gcTargetEnv: queryParameters.gcTargetEnv
+                });
+                pcEnvironment = clientApp.pcEnvironment;
+                Vue.prototype.$clientApp = clientApp;
+            } else if (pcEnvironment) {
```
Comment (justinaray) [inline examples/agentInteractionsDemo/app.js:169]: eventually, we can kill off support for pcEnvironment in the demo apps.  But, smart to leave it in for now.  Until we update all config using them.

File: examples/directoryNavigationDemo.html [hunk 5]
```diff
@@ -272,32 +279,6 @@ <h2>Groups</h2>
                 }
             }
 
-            /**
-             * Determine the embedding Genesys Cloud environment seeded on the query string or
-             * being returned through the OAuth2 Implicit grant state hash param.
-             *
-             * @returns A string indicating the embedding PC env (e.g. mypurecloud.com, mypurecloud.jp); otherwise, null.
-             */
-            function getEmbeddingPCEnv() {
```
Comment (justinaray) [inline examples/directoryNavigationDemo.html:281]: thanks for cleaning up all these duplicated utils.  Years of tech debt in the making.

File: examples/externalContactsDemo/script.js [hunk 6]
```diff
@@ -105,7 +109,7 @@ document.addEventListener('DOMContentLoaded', function () {
                         }
                     }
                 }).catch(err => {
-                    setErrorState(err);
+                    setErrorState(err.message);
```
Comment (justinaray) [inline examples/externalContactsDemo/script.js:112]: was this failing or just something you spotted?  Seems like the right change to make but tough to know in js and a catch block where err is `any`.

Dont mind the change ... just curious.

File: examples/utils/oauth.js (hunk 3 above)
Comment (justinaray) [reply examples/utils/oauth.js:2]: lgtm!

File: examples/externalContactsDemo/script.js (hunk 6 above)
Comment (cadgerfeast) [reply examples/externalContactsDemo/script.js:112]: When I tested the sample and got the permission error (missing app scope), the message shown in the progress bar was `[object Object]`, so I just made the change to show the real message sent by API.

**Comment by cadgerfeast** [comment]
Hey @justinaray ,
I fixed the documentation issue and tested every sample on `2.6.0-alpha` branch with local and dev `web directory`.
Couldn't find any issue.

**Pull request merged**